### PR TITLE
Add odra-macros crate

### LINKED ISSUE #219: Add `odra-macros` crate
Migrate `odra-macros` from the playground repo and provide a macro `odra::module2`. It should produce code that works with `mock-vm` at first.

### PR DESCRIPTION
**Pull request by kpob**

### CODE REVIEW

File: examples2/src/erc20.rs [hunk 1]
```diff
@@ -125,7 +126,7 @@ impl Erc20 {
     pub fn cross_transfer(&mut self, other: Address, to: Address, value: U256) {
         let caller = self.env().caller();
 
-        let other_erc20 = Erc20ContractRef {
+        let mut other_erc20 = Erc20ContractRef {
```
Comment (zie1ony) [inline examples2/src/erc20.rs:129]: Generated Refs, should have `new(address: Address, env: &Rc<ContractEnv>)`.

File: odra-macros/src/ir/deployer_item.rs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    use quote::quote;
+
+    #[test]
+    fn deployer_impl() {
+        let module = test_utils::mock_module();
+        let expected = quote! {
+            pub struct Erc20Deployer;
+
+            impl Erc20Deployer {
+                pub fn deploy(env: &mut Env) -> Erc20Ref {
```
Comment (zie1ony) [inline odra-macros/src/ir/deployer_item.rs:51]:
```suggestion
                pub fn deploy(env: &mut odra::HostEnv) -> super::Erc20Ref {
```

File: odra-macros/src/ir/ref_item.rs [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+            })
+        };
+
+        let return_type = fun.return_type();
+        let mutability = fun.is_mut().then(|| quote!(mut));
+        parse_quote!(
+            pub fn #fun_name(& #mutability self #typed_args) #return_type {
+                self.env.call_contract(
+                    self.address,
+                    CallDef::new(String::from(#fun_name_str), #args),
```
Comment (zie1ony) [inline odra-macros/src/ir/ref_item.rs:45]: You should use full paths to `CallDef` and `String` (from prelude).

File: odra-macros/src/ir/ref_item.rs [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+                pub fn address(&self) -> &Address {
+                    &self.address
+                }
+
+                pub fn init(&mut self, total_supply: Option<U256>) {
+                    self.env.call_contract(
+                        self.address,
+                        CallDef::new(
+                            String::from("init"),
+                            {
```
Comment (zie1ony) [inline odra-macros/src/ir/ref_item.rs:107]: This is nice :)

**Review by zie1ony — APPROVED** [review]

File: odra-macros/src/ir/ref_item.rs (hunk 3 above)
Comment (kpob) [reply odra-macros/src/ir/ref_item.rs:45]: Yes, I discovered it in the meantime, It's already fixed, but will be added  to the next pr.

File: odra-macros/src/ir/deployer_item.rs (hunk 2 above)
Comment (kpob) [reply odra-macros/src/ir/deployer_item.rs:51]: Taken from the playground, it's not a part of this PR, will be done properly in upcoming PRs.

File: examples2/src/erc20.rs (hunk 1 above)
Comment (kpob) [reply examples2/src/erc20.rs:129]: Ok, I'll fix in the next PR.

**Pull request merged**